File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/DeformableConv2D.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.layers.convolution;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.nd4j.linalg.api.ops.impl.layers.convolution.config.DeformableConv2DConfig;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Deformable Convolution 2D
 *
 * Implements deformable convolution where learned offsets are added to
 * the regular sampling grid, allowing the convolution to adapt to
 * geometric transformations in the input.
 *
 * Reference: "Deformable Convolutional Networks" (Dai et al., 2017)
 *            "Deformable ConvNets v2" (Zhu et al., 2019)
 *
 * @author Eclipse Deeplearning4j
 */
@NoArgsConstructor
@Getter
public class DeformableConv2D extends DynamicCustomOp {

    protected DeformableConv2DConfig config;

    /**
     * Create a deformable convolution 2D operation.
     *
     * @param input Input tensor [batch, channels, height, width] (NCHW) or NHWC
     * @param weights Convolution weights [out_channels, in_channels/groups, kH, kW]
     * @param offset Learned offsets [batch, 2*kH*kW*offset_groups, out_h, out_w]
     * @param config Configuration for the deformable convolution
     */
    public DeformableConv2D(@NonNull INDArray input, @NonNull INDArray weights,
                           @NonNull INDArray offset, @NonNull DeformableConv2DConfig config) {
        super(new INDArray[]{input, weights, offset}, null);
        this.config = config;
        addArgs();
    }

    /**
     * Create a deformable convolution 2D operation with bias.
     *
     * @param input Input tensor
     * @param weights Convolution weights
     * @param offset Learned offsets
     * @param bias Bias [out_channels]
     * @param config Configuration
     */
    public DeformableConv2D(@NonNull INDArray input, @NonNull INDArray weights,
                           @NonNull INDArray offset, INDArray bias,
                           @NonNull DeformableConv2DConfig config) {
        super(bias != null ?
                new INDArray[]{input, weights, offset, bias} :
                new INDArray[]{input, weights, offset}, null);
        this.config = config;
        addArgs();
    }

    /**
     * Create a deformable convolution 2D operation with bias and mask (v2).
     *
     * @param input Input tensor
     * @param weights Convolution weights
     * @param offset Learned offsets
     * @param bias Bias (optional, can be null)
     * @param mask Modulation mask for v2 (optional, can be null)
     * @param config Configuration
     */
    public DeformableConv2D(@NonNull INDArray input, @NonNull INDArray weights,
                           @NonNull INDArray offset, INDArray bias, INDArray mask,
                           @NonNull DeformableConv2DConfig config) {
        super(buildInputs(input, weights, offset, bias, mask), null);
        this.config = config;
        addArgs();
    }

    /**
     * Create a deformable convolution 2D for SameDiff.
     *
     * @param sd SameDiff instance
     * @param input Input tensor
     * @param weights Convolution weights
     * @param offset Learned offsets
     * @param config Configuration
     */
    @Builder(builderMethodName = "sameDiffBuilder")
    public DeformableConv2D(@NonNull SameDiff sd, @NonNull SDVariable input,
                           @NonNull SDVariable weights, @NonNull SDVariable offset,
                           @NonNull DeformableConv2DConfig config) {
        super(null, sd, new SDVariable[]{input, weights, offset}, false);
        this.config = config;
        addArgs();
    }

    /**
     * Create a deformable convolution 2D for SameDiff with all options.
     *
     * @param sd SameDiff instance
     * @param input Input tensor
     * @param weights Convolution weights
     * @param offset Learned offsets
     * @param bias Bias (optional)
     * @param mask Modulation mask for v2 (optional)
     * @param config Configuration
     */
    public DeformableConv2D(@NonNull SameDiff sd, @NonNull SDVariable input,
                           @NonNull SDVariable weights, @NonNull SDVariable offset,
                           SDVariable bias, SDVariable mask,
                           @NonNull DeformableConv2DConfig config) {
        super(null, sd, buildSdInputs(input, weights, offset, bias, mask), false);
        this.config = config;
        addArgs();
    }

    /**
     * Create a deformable convolution 2D for SameDiff with individual parameters (for codegen).
     */
    public DeformableConv2D(@NonNull SameDiff sd, @NonNull SDVariable input,
                           @NonNull SDVariable weights, @NonNull SDVariable offset,
                           SDVariable bias, SDVariable mask,
                           int kH, int kW, int sH, int sW, int pH, int pW,
                           int dH, int dW, int groups, int deformableGroups, boolean isNCHW) {
        super(null, sd, buildSdInputs(input, weights, offset, bias, mask), false);
        this.config = DeformableConv2DConfig.builder()
                .kH(kH).kW(kW)
                .sH(sH).sW(sW)
                .pH(pH).pW(pW)
                .dH(dH).dW(dW)
                .groups(groups)
                .deformableGroups(deformableGroups)
                .isNCHW(isNCHW)
                .build();
        addArgs();
    }

    /**
     * Create a deformable convolution 2D for INDArray with individual parameters (for codegen).
     */
    public DeformableConv2D(@NonNull INDArray input, @NonNull INDArray weights,
                           @NonNull INDArray offset, INDArray bias, INDArray mask,
                           int kH, int kW, int sH, int sW, int pH, int pW,
                           int dH, int dW, int groups, int deformableGroups, boolean isNCHW) {
        super(buildInputs(input, weights, offset, bias, mask), null);
        this.config = DeformableConv2DConfig.builder()
                .kH(kH).kW(kW)
                .sH(sH).sW(sW)
                .pH(pH).pW(pW)
                .dH(dH).dW(dW)
                .groups(groups)
                .deformableGroups(deformableGroups)
                .isNCHW(isNCHW)
                .build();
        addArgs();
    }

    private static INDArray[] buildInputs(INDArray input, INDArray weights, INDArray offset,
                                          INDArray bias, INDArray mask) {
        List<INDArray> inputs = new ArrayList<>();
        inputs.add(input);
        inputs.add(weights);
        inputs.add(offset);
        if (bias != null) {
            inputs.add(bias);
        }
        if (mask != null) {
            inputs.add(mask);
        }
        return inputs.toArray(new INDArray[0]);
    }

    private static SDVariable[] buildSdInputs(SDVariable input, SDVariable weights, SDVariable offset,
                                              SDVariable bias, SDVariable mask) {
        List<SDVariable> inputs = new ArrayList<>();
        inputs.add(input);
        inputs.add(weights);
        inputs.add(offset);
        if (bias != null) {
            inputs.add(bias);
        }
        if (mask != null) {
            inputs.add(mask);
        }
        return inputs.toArray(new SDVariable[0]);
    }

    private void addArgs() {
        addIArgument(
                config.getKH(),
                config.getKW(),
                config.getSH(),
                config.getSW(),
                config.getPH(),
                config.getPW(),
                config.getDH(),
                config.getDW(),
                config.getGroups(),
                config.getDeformableGroups(),
                config.isNCHW() ? 1 : 0
        );
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        if (iArguments.size() >= 11) {
            this.config = DeformableConv2DConfig.builder()
                    .kH(iArguments.get(0).intValue())
                    .kW(iArguments.get(1).intValue())
                    .sH(iArguments.get(2).intValue())
                    .sW(iArguments.get(3).intValue())
                    .pH(iArguments.get(4).intValue())
                    .pW(iArguments.get(5).intValue())
                    .dH(iArguments.get(6).intValue())
                    .dW(iArguments.get(7).intValue())
                    .groups(iArguments.get(8).intValue())
                    .deformableGroups(iArguments.get(9).intValue())
                    .isNCHW(iArguments.get(10) != 0)
                    .build();
        }
    }

    @Override
    public boolean isConfigProperties() {
        return true;
    }

    @Override
    public String configFieldName() {
        return "config";
    }

    @Override
    public String opName() {
        return "deformable_conv2d";
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        Preconditions.checkState(inputDataTypes != null && inputDataTypes.size() >= 3,
                "Expected at least 3 input data types for %s, got %s", getClass(), inputDataTypes);
        return Collections.singletonList(inputDataTypes.get(0));
    }

    @Override
    public String onnxName() {
        return "DeformConv";
    }
}

File: libnd4j/include/ops/declarable/generic/nn/convo/deformable_conv2d.cpp
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

//
// @author Eclipse Deeplearning4j
//
// Reference: "Deformable Convolutional Networks" (Dai et al., 2017)
//            "Deformable ConvNets v2" (Zhu et al., 2019)
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_deformable_conv2d)

#include <ops/declarable/headers/nn.h>
#include <ops/declarable/helpers/deformable_conv.h>

namespace sd {
namespace ops {

/**
 * deformable_conv2d - Deformable Convolution 2D
 *
 * Implements deformable convolution where learned offsets are added to
 * the regular sampling grid, allowing the convolution to adapt to
 * geometric transformations in the input.
 *
 * Inputs:
 *   0: input [batch, channels, height, width] (NCHW) or [batch, height, width, channels] (NHWC)
 *   1: weights [out_channels, in_channels/groups, kernel_h, kernel_w]
 *   2: offset [batch, 2*kernel_h*kernel_w*offset_groups, out_h, out_w]
 *   3: bias [out_channels] (optional)
 *   4: mask [batch, kernel_h*kernel_w*offset_groups, out_h, out_w] (optional, for v2)
 *
 * Integer args:
 *   0: kernel_h
 *   1: kernel_w
 *   2: stride_h
 *   3: stride_w
 *   4: pad_h
 *   5: pad_w
 *   6: dilation_h
 *   7: dilation_w
 *   8: groups
 *   9: offset_groups (deformable groups)
 *   10: isNCHW (1 for NCHW, 0 for NHWC)
 *
 * Outputs:
 *   0: output [batch, out_channels, out_h, out_w] or [batch, out_h, out_w, out_channels]
 */
CUSTOM_OP_IMPL(deformable_conv2d, 3, 1, false, 0, 11) {
    auto input = INPUT_VARIABLE(0);
    auto weights = INPUT_VARIABLE(1);
    auto offset = INPUT_VARIABLE(2);
    auto bias = block.width() > 3 ? INPUT_VARIABLE(3) : nullptr;
    auto mask = block.width() > 4 ? INPUT_VARIABLE(4) : nullptr;

    auto output = OUTPUT_VARIABLE(0);

    // Get integer arguments
    auto kH = INT_ARG(0);
    auto kW = INT_ARG(1);
    auto sH = INT_ARG(2);
    auto sW = INT_ARG(3);
    auto pH = INT_ARG(4);
    auto pW = INT_ARG(5);
    auto dH = INT_ARG(6);
    auto dW = INT_ARG(7);
    auto groups = INT_ARG(8);
    auto offsetGroups = INT_ARG(9);
    auto isNCHW = INT_ARG(10) != 0;

    REQUIRE_TRUE(input->rankOf() == 4, 0,
                 "deformable_conv2d: input must be rank 4, got %i", input->rankOf());
    REQUIRE_TRUE(weights->rankOf() == 4, 0,
                 "deformable_conv2d: weights must be rank 4, got %i", weights->rankOf());
    REQUIRE_TRUE(offset->rankOf() == 4, 0,
                 "deformable_conv2d: offset must be rank 4, got %i", offset->rankOf());

    sd::LongType batchSize, inChannels, inputH, inputW;
    if (isNCHW) {
        batchSize = input->sizeAt(0);
        inChannels = input->sizeAt(1);
        inputH = input->sizeAt(2);
        inputW = input->sizeAt(3);
    } else {
        batchSize = input->sizeAt(0);
        inputH = input->sizeAt(1);
        inputW = input->sizeAt(2);
        inChannels = input->sizeAt(3);
    }

    auto outChannels = weights->sizeAt(0);
    auto outputH = (inputH + 2 * pH - dH * (kH - 1) - 1) / sH + 1;
    auto outputW = (inputW + 2 * pW - dW * (kW - 1) - 1) / sW + 1;

    // Validate offset shape
    REQUIRE_TRUE(offset->sizeAt(0) == batchSize, 0,
                 "deformable_conv2d: offset batch size must match input");
    REQUIRE_TRUE(offset->sizeAt(1) == 2 * kH * kW * offsetGroups, 0,
                 "deformable_conv2d: offset channels must be 2*kH*kW*offset_groups");

    // Validate mask shape if provided
    if (mask != nullptr && !mask->isEmpty()) {
        REQUIRE_TRUE(mask->sizeAt(0) == batchSize, 0,
                     "deformable_conv2d: mask batch size must match input");
        REQUIRE_TRUE(mask->sizeAt(1) == kH * kW * offsetGroups, 0,
                     "deformable_conv2d: mask channels must be kH*kW*offset_groups");
    }

    // Handle NHWC format by permuting to NCHW for internal processing
    NDArray* inputNCHW = const_cast<NDArray*>(input);
    NDArray* outputNCHW = output;
    NDArray* inputPermuted = nullptr;
    NDArray* outputTemp = nullptr;

    if (!isNCHW) {
        // Permute input from NHWC to NCHW
        std::vector<sd::LongType> permDims = {0, 3, 1, 2};
        inputPermuted = new NDArray(input->permute(permDims, false, false));
        inputNCHW = inputPermuted;

        // Create temporary output in NCHW format
        std::vector<sd::LongType> outShape = {batchSize, outChannels, outputH, outputW};
        outputTemp = new NDArray('c', outShape, input->dataType(), block.launchContext());
        outputNCHW = outputTemp;
    }

    // Use helper for CPU/GPU implementation
    helpers::deformableConv2d(block.launchContext(),
                               inputNCHW, weights, offset, bias, mask, outputNCHW,
                               kH, kW, sH, sW, pH, pW, dH, dW, groups, offsetGroups);

    // Convert back to NHWC if needed
    if (!isNCHW) {
        std::vector<sd::LongType> permDimsBack = {0, 2, 3, 1};
        auto permutedBack = outputNCHW->permute(permDimsBack, false, false);
        output->assign(permutedBack);
        delete permutedBack;
        delete inputPermuted;
        delete outputTemp;
    }

    return sd::Status::OK;
}

DECLARE_TYPES(deformable_conv2d) {
    getOpDescriptor()->addTraits(OP_TRAIT_FULLY_WRITING);
    getOpDescriptor()->setAllowedInputTypes({ALL_FLOATS})->setSameMode(true);
}

DECLARE_SHAPE_FN(deformable_conv2d) {
    auto inShape = inputShape->at(0);
    auto weightsShape = inputShape->at(1);
    auto inputType = sd::ArrayOptions::dataType(inShape);

    auto kH = INT_ARG(0);
    auto kW = INT_ARG(1);
    auto sH = INT_ARG(2);
    auto sW = INT_ARG(3);
    auto pH = INT_ARG(4);
    auto pW = INT_ARG(5);
    auto dH = INT_ARG(6);
    auto dW = INT_ARG(7);
    auto isNCHW = INT_ARG(10) != 0;

    sd::LongType batchSize, inH, inW;
    if (isNCHW) {
        batchSize = shape::sizeAt(inShape, static_cast<sd::LongType>(0));
        inH = shape::sizeAt(inShape, static_cast<sd::LongType>(2));
        inW = shape::sizeAt(inShape, static_cast<sd::LongType>(3));
    } else {
        batchSize = shape::sizeAt(inShape, static_cast<sd::LongType>(0));
        inH = shape::sizeAt(inShape, static_cast<sd::LongType>(1));
        inW = shape::sizeAt(inShape, static_cast<sd::LongType>(2));
    }

    auto outChannels = shape::sizeAt(weightsShape, static_cast<sd::LongType>(0));
    auto outputH = (inH + 2 * pH - dH * (kH - 1) - 1) / sH + 1;
    auto outputW = (inW + 2 * pW - dW * (kW - 1) - 1) / sW + 1;

    std::vector<sd::LongType> outputShape;
    if (isNCHW) {
        outputShape = {batchSize, outChannels, outputH, outputW};
    } else {
        outputShape = {batchSize, outputH, outputW, outChannels};
    }

    auto outputShapeInfo = ConstantShapeHelper::getInstance().createShapeInfo(inputType, 'c', outputShape);
    return SHAPELIST(outputShapeInfo);
}

}  // namespace ops
}  // namespace sd

#endif
